### ISSUE TITLE
PubSub 0.39.1 Observed non-recoverable stream error 403 User not authorized to perform...

### ISSUE THREAD

**Issue by vincemartin** [googleapis/google-cloud-python#7387]
OS - Windows Server
Python version - Python 2.7.15
Pubsub version - google-cloud-pubsub==0.39.1

Code : 

def callback(message):
    logging.info('Received message {}'.format(message))
    try:
        filename = message.data
        global queue
        queue.put_nowait(filename)

        global waiting_event
        waiting_event.set()
    except Exception as e:
        logging.error("Callback : {}".format(e))
    finally:
        message.ack()

def Run(updateDatabase_Function, topic, timeout=60.0 * 60.0):
    global queue
    queue = Queue()
    global waiting_event
    waiting_event = threading.Event()
    waiting_event.clear()

    credentials = GetCredentials()
    subscriber = pubsub_v1.SubscriberClient(credentials=credentials)
    subscription_path = subscriber.subscription_path(credentials.project_id, source)
    subscriber.subscribe(subscription_path, callback=callback)

After several days of good behaviour, the client sent the following messages  :
2019-02-09 01:02:22,566 INFO : Observed non-recoverable stream error 403 User not authorized to perform this action.
2019-02-09 01:02:22,566 INFO : RPC termination has signaled streaming pull manager shutdown.
2019-02-09 01:02:22,566 INFO : Observed non-recoverable stream error 403 User not authorized to perform this action.
2019-02-09 01:02:22,566 INFO : Thread-ConsumeBidirectionalStream exiting
2019-02-09 01:02:22,628 INFO : Thread-LeaseMaintainer exiting.
2019-02-09 01:02:22,676 INFO : Thread-Heartbeater exiting.

**Comment by tseaver** [comment]
@vincemartin Is it possible that somebody removed permissions from the service account returned by your `GetCredentials`  call?  For instance, by using the Cloud Console?

@jadekler Can you suggest another reason why a working pubsub subscription stream would abort with a 403?

**Comment by jadekler** [comment]
I'm no auth expert, but could maybe be some token-refreshing problem? cc @broady who knows a lot more about auth than I do.

**Comment by vincemartin** [comment]
@tseaver : no, I don't think so. We re-launched the python script and it works again. We are just afraid to have this error coming again randomly...

**Comment by sduskis** [comment]
I've seen something like this in Java in my Cloud Bigtable days and in nodejs-pubsub (https://github.com/googleapis/nodejs-pubsub/issues/318).  Here's what I think is happening:


1. The service account credentials or compute credentials can be used to get an oauth2 token. The client caches the token, which includes an expiration timeout.
2. A bunch of requests get the oauth2 token in the header, and everything works fine
3. Something happens on the server-side to cause the cached oauth2 token to be revoked.
4. The requests that used to work don't work anymore, and the server sends UNAUTHORIZED for all requests.  The client keeps trying with the same token to no avail.

grpc-java doesn't have this problem, since it uses a client-side JWT token (see [this grpc class](https://github.com/grpc/grpc-java/blob/ce6358e57fe1ab81bd5568129beb13ab2afef129/auth/src/main/java/io/grpc/auth/GoogleAuthLibraryCallCredentials.java) and this [JWT implementation](https://github.com/googleapis/google-auth-library-java/blob/master/oauth2_http/java/com/google/auth/oauth2/ServiceAccountJwtAccessCredentials.java)).

The first step will be to write some docs around how a user can do this, and I'll also discuss a language-wide change with the Google team.

**Comment by sduskis** [comment]
@plamut the "triaged for GA" work here is to document a way to use client-side JWT.

**Comment by sduskis** [comment]
@ajaaym has an alternative explanation to the problem.  Pub/Sub subscribers are a long running streaming RPC.  It stays open while the connection is open.  At some point, usually about an hour, the connection closes, and retries kick in to open a new stream and a new connection.  

Auth tokens also stay open for about `1 hour - some buffer (5-10 minutes)` on the client side.  There's also a server-side buffer of `1 hour + 20 minutes`, just in case there are long running streams.  The Auth token should be good for 80 minutes.

If the connection restarts when the auth token is 40 minutes old, then there will be some problems.  The stream will start with an almost stale token, and will be good for another 40 minutes.  Unary RPCs would get the token refresh, and everything works fine.  Pub/Sub's long running streams would have the same token since the RPC inception, and would fail 40 minutes into the operation.

JWT also helps here. After 40 minutes, OAuth2 based tokens would fail with UNAUTHORIZED.  JWT fails  with status 2 "Stream Removed" which is retried.

In Node.js we discovered that all we need to do is not add scopes to Service Account Credentials.  That would allow grpc to convert the  Service Account Credentials to JWT.  That hopefully should work for Python as well.

**Comment by plamut** [comment]
A workaround documented in #8429 (using JWT-based authentication), removing the `triaged for GA` label

**Comment by jadekler** [comment]
> Pub/Sub's long running streams would have the same token since the RPC inception, and would fail 40 minutes into the operation.

Are we sure about that? What is the manner of failure? Who initiates the failure (the GFE...?)?

My naive understanding is that auth is only checked at RPC-init time, not periodically throughout the life of an RPC. But, I don't really understand this space super well.
